File: CS/ServerSide/MyReportStorage.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;
using DevExpress.XtraReports.UI;
using DevExpress.XtraReports.Web.Extensions;

namespace ServerSide {
    public class MyReportStorage : ReportStorageWebExtension {
        public Dictionary<string, XtraReport> Reports = new Dictionary<string, XtraReport>();

        public MyReportStorage() {
            Reports.Add("Products", new XtraReport1());
            Reports.Add("Categories", new XtraReport2());
        }
        public override bool CanSetData(string url) {
            return true;
        }
        public override byte[] GetData(string url) {
            var report = Reports[url];
            using(MemoryStream stream = new MemoryStream()) {
                report.SaveLayoutToXml(stream);
                return stream.ToArray();
            }
        }
        public override Dictionary<string, string> GetUrls() {
            return Reports.ToDictionary(x => x.Key, y => y.Key);
        }
        public override void SetData(XtraReport report, string url) {
            if(Reports.ContainsKey(url)) {
                Reports[url] = report;
            }
            else {
                Reports.Add(url, report);
            }
        }
        public override string SetNewData(XtraReport report, string defaultUrl) {
            SetData(report, defaultUrl);
            return defaultUrl;
        }
        public override bool IsValidUrl(string url) {
            return true;
        }
    }
}

File: VB/ServerSide/MyReportStorage.vb
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraReports.Web.Extensions

Namespace ServerSide
	Public Class MyReportStorage
		Inherits ReportStorageWebExtension

		Public Reports As New Dictionary(Of String, XtraReport)()

		Public Sub New()
			Reports.Add("Products", New XtraReport1())
			Reports.Add("Categories", New XtraReport2())
		End Sub
		Public Overrides Function CanSetData(ByVal url As String) As Boolean
			Return True
		End Function
		Public Overrides Function GetData(ByVal url As String) As Byte()
			Dim report = Reports(url)
			Using stream As New MemoryStream()
				report.SaveLayoutToXml(stream)
				Return stream.ToArray()
			End Using
		End Function
		Public Overrides Function GetUrls() As Dictionary(Of String, String)
			Return Reports.ToDictionary(Function(x) x.Key, Function(y) y.Key)
		End Function
		Public Overrides Sub SetData(ByVal report As XtraReport, ByVal url As String)
			If Reports.ContainsKey(url) Then
				Reports(url) = report
			Else
				Reports.Add(url, report)
			End If
		End Sub
		Public Overrides Function SetNewData(ByVal report As XtraReport, ByVal defaultUrl As String) As String
			SetData(report, defaultUrl)
			Return defaultUrl
		End Function
		Public Overrides Function IsValidUrl(ByVal url As String) As Boolean
			Return True
		End Function
	End Class
End Namespace